### PR TITLE
safety: rework

### PR DESCRIPTION
**Pull request by half-duplex**
### Description
This fixes (or works towards fixing) a bunch of issues with safety.py:
- `.safety` doesn't say what the current state is
- Uses VirusTotal API v2 (though they have no current plans to deprecate it) instead of v3
- Doesn't find URLs in messages containing anything else ("`foo http://example.com`" fails)
- Doesn't make local cache accessible to other plugins
- Doesn't use already-cached VT results in local mode
- Doesn't allow configuration of local or strict defaults
- url.py was not checking the domain blacklist and was also failing to handle "`foo http://example.com`"
- Other tweaks and improvements

### Checklist
- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: sopel/modules/url.py
Comment (Exirel) [inline sopel/modules/url.py:312]: You can simplify this with:

```suggestion
        if url in bot.memory.get("safety_cache", {}):
```

File: sopel/modules/url.py
Comment (Exirel) [inline sopel/modules/url.py:316]: Same simplification here:

```suggestion
        if netloc in bot.memory.get("safety_cache_local", {}):
```

File: sopel/modules/safety.py
Comment (Exirel) [inline sopel/modules/safety.py:122]: Since you are adding type hint, I invite you to also type `bot`: it can be either a `sopel.bot.Sopel` or a `sopel.bot.SopelWrapper`.

**Review by Exirel — CHANGES_REQUESTED** [review]

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:31]: To make combining with the b64'd domain not require bytes.decode() just for requests to str.encode()

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:34]: Ah, true. https://github.com/psf/requests/issues/2766

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:90]: I'm not sure what you're suggesting - If there was a way to use BooleanAttribute as optional I'd have done that and checked `is None` and `== False`, but if no default is specified we get `== False`. No way I saw to allow default_mode to override enabled_by_default without making the latter a pretend-bool.

Comment (Exirel) [reply sopel/modules/safety.py:90]: What I'm suggesting is to not bother removing `enabled_by_default` at all: keep it a boolean (with a default value set to `True`), and do the following:

* set `enabled_by_default`'s default value to `True`
* if `default_mode` is set, use that, and ignore `enabled_by_default`
* otherwise, if `enabled_by_default` is `True`, then set `default_mode` to `on` (or whatever appropriate value we want as the default)
* if it is set to `False`, then set `default_mode` to `off` or something similar to deactivate it

No need to transform `enabled_by_default` into a pseudo-boolean value with `None` for default, since you don't have to care about it. Silently do the migration, document it, and it'll be fine.

File: sopel/modules/safety.py
Comment (Exirel) [reply sopel/modules/safety.py:33]: Not a big fan of dealing with bytes in the code. It works, yet I don't feel very at ease with that. Oh well. Doesn't matter.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:90]: Ah, so migrate then *save*? I was going to avoid an automatic save since last I saw that clobbers comments, formatting, ordering, etc. in the config file.

Comment (Exirel) [reply sopel/modules/safety.py:90]: Nah, no need to save. When people will upgrade to Sopel 9, having set `enabled_by_default` to `False`, they will get a warning that `enabled_by_default` is not a recognized setting. Given it's a major upgrade, people should be a bit more careful. The worst that can happen here is that they get safety activated by default instead. Not a big deal in my opinion.

Comment (half-duplex) [reply sopel/modules/safety.py:90]: Ehh. I'm strongly against sending every URL the bot sees to VirusTotal by surprise, but I suppose if that's not wanted, they probably don't have a VT api key configured.
This could also be made more simple by being able to call BooleanAttribute.parse("on"), but eh.
See new push

Comment (Exirel) [reply sopel/modules/safety.py]: This new version is pretty good! And better than I suggested.

**Review by Exirel — DISMISSED** [review]

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:71]: Using "allowlist" as a verb is stylistically icky. Using it as a single word in any sense is icky, too, for that matter.

I think changes like this come across as performative, and [ignore the original derivation](https://seattlecollegian.com/op-ed-blacklist-and-whitelist-arent-racist-words-you-are/) of the terms, but that's not a good reason to block this whole PR. I'm not that petty. Just… If you're dead set on eliminating the word "whitelist" let's please at least change it to something that sounds like English. Some alternatives:

```suggestion
        "Enter any domains to trust",
```

```suggestion
        "Enter any domains to consider trusted",
```

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:112]:
```suggestion
```

One day I will get around to installing/configuring/whatever the flake8 plugin/rule that flags this as an error.

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:132]: Is the possibility of doing so really in question, or is this actually a TODO asking some future person to implement checking said header?

Checking the output of `curl --head $stevenblack_url` shows that GitHub serves an [ETag header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag), which can be saved and compared before making future requests for the full file body.

```suggestion
        # TODO: Use ETag header to avoid the download if it's unmodified
```

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:180]: This got me to check what `search_urls()` actually finds. The default list of schemes includes ftp:

https://github.com/sopel-irc/sopel/blob/2e25645487c87108330a992f68336a6b2da42188/sopel/tools/web.py#L251

Unless this plugin handles FTP links, it should search for URLs as below.

```suggestion
    for url in tools.web.search_urls(trigger, schemes=['http', 'https']):
        safe_url = "hxx" + url[3:]
```

**Review by dgw — COMMENTED** [review]
Non-blocking style/correctness nits.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:132]: Mostly a suggestion for a future improvement, since it was beyond the scope of my original unfscking goals. I guess that could be easily stuck in the DB.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:180]: VT doesn't seem to take FTP links, but I don't mind running them through the hosts blocklist. Will .startswith() it

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:71]: I'd changed the last bit in one place, but not in the other two. Why are plugin config options documented in three separate places? I understand doc phrasing vs prompt phrasing, but where's the table used?

Comment (dgw) [reply sopel/modules/safety.py:71]: The table gets slurped up by the part of the website that generates [plugin configuration info](https://sopel.chat/usage/plugin-configuration/). Is there a better way? Absolutely there is, but no one has bothered refactoring this bit yet. 🤷‍♂️

**Comment by half-duplex** [comment]
I meant to keep this as at least two separate commits, but accidentally squashed them together :sob:
Re-requesting reviews because of nontrivial logic changes and new command.

image

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:257]: Kinda gratuitous refactoring but OK.

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:174]: Missing a space? Weird to have different logger calls use "unsafedomains list" and "unsafe domain list" interchangeably.

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:175]: Probably should rename this variable if the list is going to be treated/logged/configured as "unsafe domains".

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:346]: Don't quite understand what this comment is telling me.

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:366]: I am very confused. Why replace the scheme prefix _and_ insert ZWSP between every character? Seems excessive, especially because `zwsp(thing)` as defined above will return `4 * len(thing) - 3` bytes when encoded to UTF-8 for sending to IRC (`'\u200b'.encode('utf-8')` ➡️ `b'\xe2\x80\x8b'`).

For nontrivial real-world URLs (not just the domain part) that might actually link to malware, this approach will quickly make the output of this command longer than the 510 total bytes available for one IRC line.

**Review by dgw — COMMENTED** [review]
OK, I have been nudged into "actual review" territory.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:175]: Was originally trying to not muck around inside that block much, but yeah.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:257]: It started feeling silly with `formatting.bold(formatting.color("foo bar", formatting.colors.RED))`

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:346]: If VT changes how it returns data, the structure of that dict will change, with or without a sopel version change.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:366]: The length concern is fair. The idea was that I don't know what clients use to decide what's a link or not (whether they'd catch just `example.com` out of `hxxp://example.com/`, and when quoting a possibly-malware domain we don't want it to ever be linkified.
I also didn't want to not refer to the link somehow when this may wait 10 sec for a scan & be confusing in a busy channel, though the same applies to the regular "WARNING:" lines.

```irc
<+mal> does everybody's client properly delinkify this? hxxp://example.com/
< RhinosF1> If I share it, it becomes a normal link
< RhinosF1> The protocol gets stripped
```

So, I think it's important to invalidate the link in at least one more way. I'll switch it to only ZWSP the periods.

File: sopel/modules/safety.py
Comment (dgw) [reply sopel/modules/safety.py:346]: We should have a comment like this next to every use of API data in any plugin, then.

Of course the API data format _could_ change, but it _won't_ if VT is actually serious about the API version number.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py]: No more ZWSP, just hxxp and [.]ing the .s and :s

**Comment by dgw** [comment]
I'm not going to dig into this much more. Just noting that I added another flake8 plugin to the todo list in #1765 after seeing the mess of imports that type-hinting this plugin added.

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:95]: This should contain a `TODO` so we'll find it later when looking for such things.

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:98]: Shouldn't this be a warning?

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:157]: Nittiest of nits.

```suggestion
        LOGGER.warning("Unsafe domain list download failed; using cache")
```

File: sopel/modules/safety.py
Comment (dgw) [inline sopel/modules/safety.py:179]: Unclear why this is a `.debug()` but most messages related to the domain list fetch/load process are `.info()`s.

**Review by dgw — COMMENTED** [review]
I actually thought this was ready to go, but apparently not. Now that I've gone through it once more it's time to find out why GH still shows it in my "Approved 8.0.0" view, and make it not do that.

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:95]: also added one on the malwaredomains migration

File: sopel/modules/safety.py
Comment (half-duplex) [reply sopel/modules/safety.py:179]: Changed a few, since there's a higher chance of caring about an actual update than it being loaded. Also rebased for merge conflict.

File: sopel/modules/url.py
Comment (dgw) [inline sopel/modules/url.py:322]: Why the no-op quote change? It's not even consistent with the above rewrite.

**Review by dgw — APPROVED** [review]
All good from my perspective. Probably still wait for re-review from @Exirel to be sure everything he requested is satisfactory too.

File: sopel/modules/url.py
Comment (half-duplex) [reply sopel/modules/url.py]: Huh. Don't know. I may have gotten confused in trying to keep the diff tight? Fixed

File: sopel/modules/url.py
Comment (Exirel) [inline sopel/modules/url.py:311]: Each `dict.get` or `dict[...]` call cost a little more than doing this:

```suggestion
    urls = []
    safety_cache = bot.memory.get("safety_cache", {})
    safety_cache_local = bot.memory.get("safety_cache_local", {})
    for url in unchecked_urls:
        # Avoid fetching known malicious links
        if url in safety_cache and safety_cache[url]["positives"] > 0:
            continue
        hostname = urlparse(url).hostname.lower()
        if hostname in safety_cache_local:
            continue
```

And yes, it could be a list-comprehension. Or put that in a function that would `yield url`... so many options. A for-loop is fine.

**Review by Exirel — CHANGES_REQUESTED** [review]
While we are upgrading the `safety` plugin, I think we should replace all the `"safety_cache"` hard-coded key into a constant, so `bot.memory["satefy_cache"]` would become `bot.memory[SAFETY_CACHE_KEY]`.

They are also some probable nitpick around docstrings that lack a period here and there, but I couldn't be bothered to check them all.

**Comment by half-duplex** [comment]
Should url.py import safety.SAFETY_CACHE_KEY, or just use the string?
Also considering whether I should just find a way to merge safety_cache and safety_cache_local...

**Comment by Exirel** [comment]
> Should url.py import safety.SAFETY_CACHE_KEY, or just use the string?

You can import it if you want, both option are fine by me. We can accept that URL totally depends on the safety plugin, because it kind of does. As long as safety is part of core, anyway. Which shouldn't change yet.

**Comment by half-duplex** [comment]
I don't think URL *does* depend on the safety plugin - if safety is absent, url just skips those checks. Importing the constant would require a conditional to keep the dependency separate.

**Comment by half-duplex** [comment]
Ok, if something uses safety.py's key without following the format there will be issues, but url.py doesn't fail to perform any of its  functions if safety.py is absent. I can add two more .get()s if you want, but either way I think I prefer using the string.

**Comment by dgw** [comment]
If the code only even tries to read the safety info if the safety plugin is available, you minimize the chance of stepping on something else that has coopted its key name. That, to me, would be the main advantage of e.g.

```py
try:
    from sopel.modules.safety import SAFETY_CACHE_KEY
except ImportError:
    SAFETY_CACHE_KEY = None

# ... other code ...

# in the relevant function
    if SAFETY_CACHE_KEY is not None:
        # use it
```

Is that perfect? No, because `safety` might not be enabled. But there's also `bot.has_plugin('safety')` for additional sanity checking if you want that.

**Comment by Exirel** [comment]
> both option are fine by me

I'm still OK with that, as I don't think it matters too much at the moment. The Safety feature of Sopel would probably be better with a proper interface dedicated to that, instead of a Plugin. So for now, a string is fine by me. I'm just glad we could talk a bit about it, even tho it's nothing more than nitpicking at this point.

**Review by Exirel — APPROVED** [review]

**Pull request merged**
